tonic/src/transport/endpoint.rs

```diff
@@ -252,3 +203,96 @@ impl fmt::Debug for Endpoint {
         f.debug_struct("Endpoint").finish()
     }
 }
+
+/// Configures TLS settings for endpoints.
+#[cfg(feature = "tls")]
+pub struct ClientTlsConfig {
+    provider: TlsProvider,
+    domain: Option<String>,
+    cert: Option<Certificate>,
+    #[cfg(feature = "openssl")]
+    openssl_raw: Option<openssl1::ssl::SslConnector>,
+    #[cfg(feature = "rustls")]
+    rustls_raw: Option<tokio_rustls::rustls::ClientConfig>,
+}
+
+#[cfg(feature = "tls")]
+impl fmt::Debug for ClientTlsConfig {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("ClientTlsConfig")
+            .field("provider", &self.provider)
+            .finish()
+    }
+}
+
+#[cfg(feature = "tls")]
+impl ClientTlsConfig {
+    /// Creates a new `ClientTlsConfig` backed by the specified provider. Enable the `openssl` or
+    /// `rustls` features of the `tonic` crate to use OpenSSL or Rustls respectively.
+    pub fn new(provider: TlsProvider) -> Self {
```

Comment:
what do you think about providing constructors like so? This also allows us to make the enum private thus allowing us to add new variants without breaking code!

```rust
#[cfg(feature = "openssl")]
pub fn with_openssl() -> Self {...}

#[cfg(feature = "rustls")]
pub fn with_rustls() -> Self {...}
```

Reply:
This definitely seems better and leaves fewer types exposed.

Reply:
I've implemented this and made `TlsProvider` private to the crate - it's a lot more ergonomic and allows us to remove the `None` variant too.